### PR TITLE
Rework of Locations structs

### PR DESCRIPTION
**Pull request by sungam3r**

### CODE REVIEW

**Comment by Shane32** [comment]
Haha you beat me to this one. If we can merge the other PR first it will help prevent merge conflicts when merging master into develop.

**Comment by Shane32** [comment]
This is what I don't like about the labeler:

image

It doesn't know what it's doing, and then it keeps doing it over and over.

File: src/GraphQL/Language/AST/SourceLocation.cs [hunk 1]
```diff
@@ -1,10 +1,14 @@
...
+
 namespace GraphQL.Language.AST
 {
     /// <summary>
     /// Provides information regarding the location of a node within a document's original query text.
     /// </summary>
-    public class SourceLocation
+    public readonly struct SourceLocation : IEquatable<SourceLocation>
     {
+        public static readonly SourceLocation Empty = new SourceLocation(-1, -1, -1, -1);
```
Comment (Shane32) [inline src/GraphQL/Language/AST/SourceLocation.cs:10]: Missing xml comment.

File: src/GraphQL/Language/AST/SourceLocation.cs [hunk 2]
```diff
@@ -42,22 +46,13 @@ public SourceLocation(int line, int column, int start = -1, int end = -1)
...
-                return true;
-            if (obj.GetType() != GetType())
-                return false;
-            return Equals((SourceLocation)obj);
-        }
+        public static bool operator ==(SourceLocation first, SourceLocation second) => first.Equals(second);
+
+        public static bool operator !=(SourceLocation first, SourceLocation second) => !first.Equals(second);
+
+        public override bool Equals(object obj) => obj is SourceLocation loc && Equals(loc);
```
Comment (Shane32) [inline src/GraphQL/Language/AST/SourceLocation.cs:55]: Missing xml comment.

**Review by Shane32 — CHANGES_REQUESTED** [review]

**Comment by Shane32** [comment]
Structs already have comparison operators and stuff based on contents.  So if we can compare based on all fields, we should be able to remove `IEquatable` and both `Equals` and both operators.

Also see this code:

```cs
    public static class NodeExtensions
    {
        public static T WithLocation<T>(this T node, int line, int column, int start = -1, int end = -1)
            where T : AbstractNode
        {
            node.SourceLocation = new SourceLocation(line, column, start, end);
            return node;
        }
    }
```

There is only one reference to that code.  It is here:

```cs
    public static class AstNodeExtensions
    {
        public static T WithLocation<T>(this T node, ASTNode astNode)
            where T : AbstractNode
        {
            return node.WithLocation(0, 0, astNode?.Location.Start ?? -1, astNode?.Location.End ?? -1);
        }
    }
```

There are no other references to the `SourceLocation` constructor.  There are no other references besides the constructor that _write_ to the `Line` property.  It seems like we can remove `Line` and `Column`, unless I missed something.

And if `Line` and `Column` are always 0, then it is unlikely that any code compares two `SourceLocation` instances to each other.

**Comment by sungam3r** [comment]
> It doesn't know what it's doing, and then it keeps doing it over and over.

It does exactly what it was told:
```
"test":
  - any: ['src/GraphQL.ApiTests/**/*', 'src/GraphQL.DataLoader.Tests/**/*', 'src/GraphQL.Dummy/**/*', 'src/GraphQL.Harness.Tests/**/*', 'src/GraphQL.Tests/**/*', '!src/GraphQL.ApiTests/**/*.txt']
```

Let's decide if the changes in `docs2/site/docs/migrations/migration4.md` belong to the `documentation` label.

**Comment by sungam3r** [comment]
> Structs already have comparison operators and stuff based on contents. So if we can compare based on all fields, we should be able to remove IEquatable and both Equals and both operators.

Yes, but then it will be a comparison based on reflection. It is always recommended to define your own comparison methods and operators.

File: src/GraphQL/Language/AST/SourceLocation.cs (hunk 1 above)
Comment (sungam3r) [reply src/GraphQL/Language/AST/SourceLocation.cs:13]: done

File: src/GraphQL/Language/AST/SourceLocation.cs (hunk 2 above)
Comment (sungam3r) [reply src/GraphQL/Language/AST/SourceLocation.cs:65]: done

**Comment by sungam3r** [comment]
> There are no other references to the SourceLocation constructor. There are no other references besides the constructor that write to the Line property. It seems like we can remove Line and Column, unless I missed something. And if Line and Column are always 0, then it is unlikely that any code compares two SourceLocation instances to each other.

@joemcbride Could you please help?

File: src/GraphQL/Validation/ValidationError.cs [hunk 3]
```diff
@@ -42,7 +42,7 @@ public ValidationError(string originalQuery, string number, string message, para
             nodes?.Apply(n =>
             {
                 _nodes.Add(n);
-                if (n.SourceLocation != null)
+                if (n.SourceLocation != SourceLocation.Empty)
```
Comment (Shane32) [inline src/GraphQL/Validation/ValidationError.cs:45]: Are you sure this is correct? An uninitialized `SourceLocation` property will have its values set to 0, not to `SourceLocation.Empty`.

File: src/GraphQL/Validation/ValidationError.cs [hunk 4]
```diff
@@ -42,7 +42,7 @@ public ValidationError(string originalQuery, string number, string message, para
...
                 {
                     var location = new Location(new Source(originalQuery), n.SourceLocation.Start);
```
Comment (Shane32) [inline src/GraphQL/Validation/ValidationError.cs:47]: This line of code calls `GraphQLParser` to retrieve the `Line` and `Column` from the query based on the start position within the document.

File: src/GraphQL/Validation/ValidationError.cs [hunk 5]
```diff
@@ -42,7 +42,7 @@ public ValidationError(string originalQuery, string number, string message, para
...
                     AddLocation(location.Line, location.Column);
```
Comment (Shane32) [inline src/GraphQL/Validation/ValidationError.cs:48]: And here it adds the line/column to the error, without ever referencing the `SourceLocation.Line` or `SourceLocation.Column` property

**Comment by Shane32** [comment]
If you remove the `if (n.SourceLocation != null)` check (which isn't valid as a struct now), and then make these changes:

image

You will find that all tests pass.  None of that is used anywhere in the codebase.

If we want to preserve the `if (n.SourceLocation != null)` check, perhaps we could add a boolean property called `IsSet` or something.

File: src/GraphQL/Validation/ValidationError.cs (hunk 3 above)
Comment (sungam3r) [reply src/GraphQL/Validation/ValidationError.cs:45]: 🤦

**Comment by Shane32** [comment]
> Yes, but then it will be a comparison based on reflection. 

Not always.  https://medium.com/@semuserable/c-journey-into-struct-equality-comparison-deep-dive-9693f74562f1

> It is always recommended to define your own comparison methods and operators.

I agree, but let's not bog down the API with operators that are never used. The default comparison is perfectly adequate for running tests -- if there even were any!

**Comment by sungam3r** [comment]
I realized very interesting things. I'll change everything.

**Comment by sungam3r** [comment]
@Shane32 I want to release https://github.com/graphql-dotnet/parser/pull/97 first and go on. My idea is to use `GraphQLLocation` struct from parser repo directly and remove `SourceLocation`.

**Comment by Shane32** [comment]
@sungam3r The entire `Language.AST` namespace should be removed, or in some fashion combined with the GraphQLParser types.  Re-using a single struct containing a couple integers will not save any CPU cycles, as it has to copy the containing variables anyway.

If the idea is to allow third-party parsers, then I certainly wouldn't use any classes/structs from GraphQLParser, or else the third party parser would require a dependency on GraphQLParser.

I would suggest that we add another project (separate repo) with the AST classes and structs, and then both GraphQL.NET and GraphQLParser can use it via nuget.

**Comment by sungam3r** [comment]
Here the matter is completely different. I remember that @joemcbride told it happened so historically.

**Comment by Shane32** [comment]
> My idea is to use GraphQLLocation struct from parser repo directly and remove SourceLocation.

I do not think that should be done.

**Comment by sungam3r** [comment]
> I do not think that should be done.

Two structs are the same 100% binary and semantically.

File: src/GraphQL/Validation/ValidationError.cs [hunk 6]
```diff
@@ -42,11 +42,8 @@ public ValidationError(string originalQuery, string number, string message, para
             nodes?.Apply(n =>
             {
                 _nodes.Add(n);
-                if (n.SourceLocation != null)
-                {
-                    var location = new Location(new Source(originalQuery), n.SourceLocation.Start);
```
Comment (sungam3r) [inline src/GraphQL/Validation/ValidationError.cs:47]: no more memory allocation for `Source`

File: src/GraphQL/Language/CoreToVanillaConverter.cs [hunk 7]
```diff
@@ -13,18 +12,12 @@ namespace GraphQL.Language
 {
     public class CoreToVanillaConverter
     {
-        private readonly ISource _body;
```
Comment (sungam3r) [inline src/GraphQL/Language/CoreToVanillaConverter.cs:16]: I was surprised when I saw that this field is not being used. The class is stateless and can generally be made static. I declared a static instance. Maybe make the class static?

File: src/GraphQL/Execution/ExecutionError.cs [hunk 8]
```diff
@@ -139,20 +148,15 @@ public static class ExecutionErrorExtensions
...
+            where TError: ExecutionError
         {
-            if (abstractNode == null)
-                return;
+            if (abstractNode == null || document == null)
+                return error;
 
-            if (document != null)
-            {
-                var location = new Location(new Source(document.OriginalQuery), abstractNode.SourceLocation.Start);
```
Comment (sungam3r) [inline src/GraphQL/Execution/ExecutionError.cs:149]: no more memory alocation for `Source`

**Comment by sungam3r** [comment]
> The entire Language.AST namespace should be removed, or in some fashion combined with the GraphQLParser types. Re-using a single struct containing a couple integers will not save any CPU cycles, as it has to copy the containing variables anyway.

I'm not against. This has not been my goal yet.

**Comment by Shane32** [comment]
> > I do not think that should be done.
> 
> Two structs are the same 100% binary and semantically.

I understand.  But now another parser needs a reference to GraphQLParser in order to compile.  Please undo the change and I'll review again.

File: docs2/site/docs/migrations/migration4.md [hunk 9]
```diff
@@ -12,11 +12,16 @@
...
-* `IGraphType.CollectTypes` has been removed.
-* `ExecutionHelper.SubFieldsFor` has been removed.
-* `TypeCollectionContext` is now internal, also all methods with this parameter in `GraphTypesLookup` are private.
-* `GraphQLTypeReference` is now internal, also `GraphTypesLookup.ApplyTypeReferences` is now private.
+* `LightweightCache.First` method has been removed.
+* `IGraphType.CollectTypes` method has been removed.
+* `ExecutionHelper.SubFieldsFor` method has been removed.
+* `SourceLocation`, `NodeExtensions`, `AstNodeExtensions` classes have been removed.
+* `ErrorLocation` struct became `readonly`.
+* `GraphQL.Language.AST.INode.SourceLocation` property now has `GraphQLParser.AST.GraphQLLocation` type.
```
Comment (Shane32) [inline docs2/site/docs/migrations/migration4.md:20]: > `GraphQL.Language.AST.INode.SourceLocation` property now has `GraphQLParser.AST.GraphQLLocation` type.

Please undo.

**Review by Shane32 — CHANGES_REQUESTED** [review]

File: docs2/site/docs/migrations/migration4.md (hunk 9 above)
Comment (sungam3r) [reply docs2/site/docs/migrations/migration4.md]: done

File: src/GraphQL/Language/CoreToVanillaConverter.cs (hunk 7 above)
Comment (sungam3r) [reply src/GraphQL/Language/CoreToVanillaConverter.cs:16]: done

File: src/GraphQL/Language/SourceLocation.cs [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+        /// </summary>
+        public int End { get; }
+
+        public bool Equals(SourceLocation other) => Start == other.Start && End == other.End;
+
+        public override bool Equals(object obj) => obj is SourceLocation l && Equals(l);
+
+        public override int GetHashCode() => (Start, End).GetHashCode();
+
+        public override string ToString() => $"({Start},{End})";
```
Comment (Shane32) [inline src/GraphQL/Language/SourceLocation.cs:42]:
```suggestion
        /// <includedoc/>
        public override string ToString() => $"({Start},{End})";
```

**Review by Shane32 — CHANGES_REQUESTED** [review]

File: src/GraphQL/Language/SourceLocation.cs [hunk 11]
```diff
@@ -0,0 +1,44 @@
...
+        ///     { field { subfield } }
+        ///                         ^ field.Location.End = 20
+        /// </summary>
+        public int End { get; }
+
+        public bool Equals(SourceLocation other) => Start == other.Start && End == other.End;
+
+        public override bool Equals(object obj) => obj is SourceLocation l && Equals(l);
+
+        public override int GetHashCode() => (Start, End).GetHashCode();
```
[the remark this answers is not in the document]
Comment (sungam3r) [reply src/GraphQL/Language/SourceLocation.cs:40]: Can you show a link to the default implementation? I admit that there are optimized versions of the standard Equals and GetHashCode methods, but you shouldn't rely on them, as even a small code change can turn them off.

File: src/GraphQL/Language/SourceLocation.cs [hunk 12]
```diff
@@ -0,0 +1,44 @@
...
+        ///
+        /// For example,
+        ///     { field { subfield } }
+        ///                         ^ field.Location.End = 20
+        /// </summary>
+        public int End { get; }
+
+        public bool Equals(SourceLocation other) => Start == other.Start && End == other.End;
+
+        public override bool Equals(object obj) => obj is SourceLocation l && Equals(l);
```
[the remark this answers is not in the document]
Comment (sungam3r) [reply src/GraphQL/Language/SourceLocation.cs:38]: `inheritdoc` maybe?

[the remark this answers is not in the document]
Comment (sungam3r) [reply src/GraphQL/Language/SourceLocation.cs]: done with `inheritdoc`

File: src/GraphQL/Language/SourceLocation.cs [hunk 13]
```diff
@@ -0,0 +1,44 @@
...
+        /// <summary>
+        /// The index for the character immediately after the node in the source (i.e. it's exclusive).
+        ///
+        /// For example,
+        ///     { field { subfield } }
+        ///                         ^ field.Location.End = 20
+        /// </summary>
+        public int End { get; }
+
+        public bool Equals(SourceLocation other) => Start == other.Start && End == other.End;
```
[the remark this answers is not in the document]
Comment (sungam3r) [reply src/GraphQL/Language/SourceLocation.cs]: done

File: src/GraphQL/Language/SourceLocation.cs [hunk 14]
```diff
@@ -0,0 +1,44 @@
...
+        public int Start { get; }
+
+        /// <summary>
+        /// The index for the character immediately after the node in the source (i.e. it's exclusive).
+        ///
+        /// For example,
+        ///     { field { subfield } }
+        ///                         ^ field.Location.End = 20
+        /// </summary>
+        public int End { get; }
```
[the remark this answers is not in the document]
Comment (sungam3r) [reply src/GraphQL/Language/SourceLocation.cs]: Done with formatting:
изображение

File: src/GraphQL/Language/SourceLocation.cs (hunk 10 above)
Comment (sungam3r) [reply src/GraphQL/Language/SourceLocation.cs]: done

File: src/GraphQL/Language/SourceLocation.cs (hunk 11 above)
[the remark this answers is not in the document]
Comment (Shane32) [reply src/GraphQL/Language/SourceLocation.cs]: What I mean is that SourceLocation is a struct and GetHashCode would return a hash of the members.  The function you defined creates a struct and executes GetHashCode on it.  So it does the exact same thing, regardless of the .NET implementation of GetHashCode for structs.

[the remark this answers is not in the document]
Comment (Shane32) [reply src/GraphQL/Language/SourceLocation.cs]: Isn't it ValueType.GetHashCode either way?  Maybe I'm all wrong.

File: src/GraphQL/Language/AST/SourceLocation.cs [hunk 15]
```diff
@@ -17,28 +20,38 @@ public SourceLocation(int start, int end)
...
+        /// For example:
+        /// <code>
+        /// { field { subfield } }
+        /// <br/>
+        /// --------------------^ field.Location.End = 20
+        /// </code>
         /// </summary>
         public int End { get; }
 
+        /// <inheritdoc/>
```
Comment (Shane32) [inline src/GraphQL/Language/AST/SourceLocation.cs:45]: This is not inheriting anything.  Should cref the default equality operator.

**Review by Shane32 — CHANGES_REQUESTED** [review]

File: src/GraphQL/Language/AST/SourceLocation.cs (hunk 15 above)
Comment (sungam3r) [reply src/GraphQL/Language/AST/SourceLocation.cs:45]: изображение

and

изображение

Looks pretty good.

Comment (Shane32) [reply src/GraphQL/Language/AST/SourceLocation.cs:45]: Ok then if it works.  I don't understand what it's inheriting.  I think it's still wrong.

**Review by Shane32 — APPROVED** [review]

File: src/GraphQL/Language/SourceLocation.cs (hunk 11 above)
[the remark this answers is not in the document]
Comment (sungam3r) [reply src/GraphQL/Language/SourceLocation.cs]:
```c#
 /// <summary>Calculates the hash code for the current <see cref="T:System.ValueTuple`2" /> instance.</summary>
    /// <returns>The hash code for the current <see cref="T:System.ValueTuple`2" /> instance.</returns>
    public override int GetHashCode()
    {
      return ValueTuple.CombineHashCodes(EqualityComparer<T1>.Default.GetHashCode(this.Item1), EqualityComparer<T2>.Default.GetHashCode(this.Item2));
    }
```

[the remark this answers is not in the document]
Comment (sungam3r) [reply src/GraphQL/Language/SourceLocation.cs]:
```c#
using System;

namespace ConsoleApp8
{
    class Program
    {
        static void Main(string[] args)
        {
            var a = GC.GetAllocatedBytesForCurrentThread();

            var loc1 = new Location1(10, 20);

            for (int i = 0; i < 1000000; ++i)
                _ = loc1.GetHashCode();

            var b = GC.GetAllocatedBytesForCurrentThread();

            Console.WriteLine(b - a);

            var loc2 = new Location2(10, 20);

            for (int i = 0; i < 1000000; ++i)
                _ = loc2.GetHashCode();

            var c = GC.GetAllocatedBytesForCurrentThread();

            Console.WriteLine(c - b);
            Console.ReadKey();
        }
    }

    public readonly struct Location1
    {
        public Location1(int start, int end)
        {
            Start = start;
            End = end;
        }

        public override int GetHashCode() => (Start, End).GetHashCode();

        public int Start { get; }

        public int End { get; }
    }

    public readonly struct Location2
    {
        public Location2(int start, int end)
        {
            Start = start;
            End = end;
        }

        public int Start { get; }

        public int End { get; }
    }
}
```
изображение

So...

File: src/GraphQL/Language/AST/SourceLocation.cs (hunk 15 above)
Comment (sungam3r) [reply src/GraphQL/Language/AST/SourceLocation.cs:45]: It inherits docs from interface:
изображение

**Pull request merged**